gprofiler/utils/node.py

```diff
@@ -0,0 +1,176 @@
+import json
+import os
+import shutil
+import signal
+import stat
+from typing import Dict, List
+
+import psutil
+import requests
+from granulate_utils.linux.ns import get_proc_root_path, get_process_nspid, resolve_proc_root_links, run_in_ns
+from granulate_utils.linux.process import is_musl
+from retry import retry
+from websocket import create_connection
+from websocket._core import WebSocket
+
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata import get_exe_version
+from gprofiler.utils import add_permission_dir, pgrep_exe, resource_path
+
+logger = get_logger_adapter(__name__)
+
+DSO_GIT_REVISION = "20eb88a"
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def get_dest_inside_container(musl: bool, node_version: str) -> str:
+    libc = "musl" if musl else "glibc"
+    return os.path.join("/", "tmp", "node_module", DSO_GIT_REVISION, libc, node_version)
+
+
+def _start_debugger(pid: int) -> None:
+    # for windows: in shell node -e "process._debugProcess(PID)"
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
```

Comment:
Btw if you already found the source code in nodejs which performs this behavior (e.g the constant port, ..) it's nice a put a permalink code comment here to the github sources :) I'd love to take a look